objc2/src/foundation/mutable_dictionary.rs

```diff
@@ -0,0 +1,408 @@
+        /// let array = NSMutableDictionary::into_values_array(dict);
+        /// println!("{:?}", array);
+        /// ```
+        pub fn into_values_array(dict: Id<Self, Owned>) -> Id<NSArray<V, Owned>, Shared> {
+            unsafe { msg_send_id![&dict, allValues] }
+        }
+    }
+);
+
+unsafe impl<K: Message, V: Message> NSFastEnumeration for NSMutableDictionary<K, V> {
+    type Item = K;
+}
+
+impl<'a, K: Message, V: Message> Index<&'a K> for NSMutableDictionary<K, V> {
+    type Output = V;
+
+    fn index<'s>(&'s self, index: &'a K) -> &'s V {
+        self.get(index).unwrap()
+    }
+}
+
+impl<'a, K: Message, V: Message> IndexMut<&'a K> for NSMutableDictionary<K, V> {
+    fn index_mut<'s>(&'s mut self, index: &'a K) -> &'s mut V {
+        self.get_mut(index).unwrap()
+    }
+}
+
+impl<K: Message, V: Message> DefaultId for NSMutableDictionary<K, V> {
+    type Ownership = Owned;
+
+    #[inline]
+    fn default_id() -> Id<Self, Self::Ownership> {
+        Self::new()
+    }
+}
+
+impl<K: fmt::Debug + Message, V: fmt::Debug + Message> fmt::Debug for NSMutableDictionary<K, V> {
+    #[inline]
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        fmt::Debug::fmt(&**self, f)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use alloc::vec;
+
+    use crate::{
+        foundation::{NSNumber, NSString},
+        rc::{RcTestObject, ThreadTestData},
+    };
+
+    fn sample_dict() -> Id<NSMutableDictionary<NSNumber, NSObject>, Owned> {
+        NSMutableDictionary::from_keys_and_objects(
+            &[
+                &*NSNumber::new_i32(1),
+                &*NSNumber::new_i32(2),
+                &*NSNumber::new_i32(3),
+            ],
+            vec![NSObject::new(), NSObject::new(), NSObject::new()],
+        )
+    }
+
+    #[test]
+    fn test_new() {
+        let dict = NSMutableDictionary::<NSString, NSObject>::new();
+        assert!(dict.is_empty());
+    }
+
+    #[test]
+    fn test_get_mut() {
+        let mut dict = sample_dict();
+        assert!(dict.get_mut(&NSNumber::new_i32(1)).is_some());
+        assert!(dict.get_mut(&NSNumber::new_i32(2)).is_some());
+        assert!(dict.get_mut(&NSNumber::new_i32(4)).is_none());
+    }
+
+    #[test]
+    fn test_values_mut() {
+        let mut dict = sample_dict();
+        let vec = dict.values_mut();
+        assert_eq!(vec.len(), 3);
+    }
+
+    #[test]
+    fn test_insert() {
+        let mut dict = NSMutableDictionary::new();
+
+        assert!(dict
+            .insert(NSString::from_str("one"), NSObject::new())
+            .is_none());
+        assert!(dict
+            .insert(NSString::from_str("two"), NSObject::new())
+            .is_none());
+        assert!(dict
+            .insert(NSString::from_str("three"), NSObject::new())
+            .is_none());
+        assert!(dict
+            .insert(NSString::from_str("one"), NSObject::new())
+            .is_some());
```

Comment:
Nit: Formatting seems wrong here?

Reply:
It definitely looks weird, but that's how `rustfmt` decided to format each `assert!`. I changed the dictionary to use `NSNumber` for the keys so that each `assert!` is short enough to fit on a single line.